Certificate/Requester.php

```diff
@@ -0,0 +1,117 @@
+<?php
+
+namespace AcmePhp\Bundle\Certificate;
+
+use AcmePhp\Bundle\Certificate\KeyPair\DomainKeyPairProviderFactory;
+use AcmePhp\Bundle\Certificate\Storage\CertificateStorageFactory;
+use AcmePhp\Bundle\DomainConfiguration\DomainConfiguration;
+use AcmePhp\Core\AcmeClient;
+use Psr\Log\LoggerAwareTrait;
+use Psr\Log\NullLogger;
+
+class Requester
+{
+    use LoggerAwareTrait;
+
+    /**
+     * @var AcmeClient
+     */
+    protected $client;
+
+    /**
+     * @var CertificateStorageFactory
+     */
+    protected $storageFactory;
+
+    /**
+     * @var DomainKeyPairProviderFactory
+     */
+    protected $keyPairFactory;
+
+    /**
+     * Requester constructor.
+     *
+     * @param AcmeClient $client
+     */
+    public function __construct(
+        AcmeClient $client,
+        DomainKeyPairProviderFactory $keyPairFactory,
+        CertificateStorageFactory $storageFactory
+    )
+    {
+        $this->client = $client;
+        $this->keyPairFactory = $keyPairFactory;
+        $this->storageFactory = $storageFactory;
+
+        $this->logger = new NullLogger();
+    }
+
+    public function requestOrRenewCertificate(DomainConfiguration $configuration)
+    {
+        $domain = $configuration->getDomain();
+        $certificateStorage = $this->storageFactory->createCertificateStorage($domain);
+        $keyPairProvider = $this->keyPairFactory->createKeyPairProvider($domain);
+
+        if (!$keyPairProvider->hasKeyPair()) {
+            return $this->requestCertificate($configuration);
+        }
+
+        if (!$certificateStorage->hasCertificate() || !$certificateStorage->hasMetadata()) {
+            return $this->requestCertificate($configuration);
+        }
+
+        // do not use strict comparison given the first part is a unserialize object
+        if ($certificateStorage->loadMetadata() != $configuration->getCSR()) {
+            return $this->requestCertificate($configuration);
+        }
+
+        return $this->renewCertificate($configuration);
+    }
+
+    public function requestCertificate(DomainConfiguration $configuration)
+    {
+        $domain = $configuration->getDomain();
+        $certificateStorage = $this->storageFactory->createCertificateStorage($domain);
+
+        $challenge = $this->client->requestChallenge($domain);
+        $this->client->checkChallenge($challenge);
+
+        $certificateStorage->clear();
+
+        $certificate = $this->performRequest($configuration);
+
+        $this->logger->warning('Certificate for domain {domain} registered', ['domain' => $domain]);
```

Comment:
In other libraries, I use `notice` level as warning is something wrong. I think we should do the same here.